primitives/account/src/account/staking_contract/staker.rs

```diff
@@ -276,30 +276,52 @@ impl StakingContract {
 
         // All checks passed, not allowed to fail from here on!
 
-        // Create logs.
-        tx_logger.push_log(Log::UpdateStaker {
-            staker_address: staker_address.clone(),
-            old_validator_address: staker.delegation.clone(),
-            new_validator_address: delegation.clone(),
-        });
-
         // Create the receipt.
         let receipt = StakerReceipt {
             delegation: staker.delegation.clone(),
+            active_balance: staker.balance,
+            inactive_release: staker.inactive_release,
+            reactivate_all_stake,
         };
 
-        // We allow updates only when the balance is zero (the staker has been removed already)
-        // or the delegation is None. Thus, we do not need to update the source validator.
+        // Store old information for the log
+        let old_validator_address = staker.delegation.clone();
+
+        // We allow updates only when the balance is zero (the staker's stake has been removed already)
+        // or the delegation is None. Thus, we only need to update the validator's stakers counter.
+
+        // If we were delegating to a validator, we remove ourselves from it.
+        if let Some(validator_address) = &staker.delegation {
+            self.unregister_staker_from_validator(store, validator_address);
+        }
 
         // Update the staker's delegation.
-        staker.delegation = delegation;
+        staker.delegation = new_delegation;
 
-        // If we are now delegating to a validator and have a non-zero balance, we add ourselves to it.
-        if staker.delegation.is_some() && !staker.balance.is_zero() {
-            self.add_staker_to_validator(store, &staker)
-                .expect("inconsistent contract state");
+        if reactivate_all_stake {
+            // Update the staker's balance.
+            staker.balance += staker.inactive_balance;
+            staker.inactive_balance = Coin::ZERO;
+            staker.inactive_release = None;
         }
 
+        // If we are now delegating to a validator we add ourselves to it.
+        if let Some(validator_address) = &staker.delegation {
+            self.register_staker_on_validator(store, validator_address, false);
+            if reactivate_all_stake {
+                self.increase_stake_to_validator(store, validator_address, staker.balance);
+            }
+        }
```

Comment:
I believe `increase_stake_to_validator` must also be called when `reactivate_all_stake` is false, but the active stake is not zero. That can happen when delegation was `None`.

```diff
-             if reactivate_all_stake {
+             if !staker.balance.is_zero() {
```

Technically, the condition is not even needed and `increase_stake_to_validator` can just always be called, even when `staker.balance` is `0`.

Add a test where delegation is None but active balance is something, then update delegation to Some, then check the validator balance.

Reply:
Great catch!